builder/proxmox/common/step_finalize_config.go

```diff
@@ -17,18 +17,19 @@ import (
 // stepFinalizeTemplateConfig does any required modifications to the configuration _after_
 // the VM has been converted into a template, such as updating name and description, or
 // unmounting the installation ISO.
-type stepFinalizeTemplateConfig struct{}
+type stepFinalizeConfig struct{}
 
-type templateFinalizer interface {
+type finalizer interface {
 	GetVmConfig(*proxmox.VmRef) (map[string]interface{}, error)
 	SetVmConfig(*proxmox.VmRef, map[string]interface{}) (interface{}, error)
+	StartVm(*proxmox.VmRef) (string, error)
```

Comment:
Is there a reason to add `StartVm` to that contract? It seems that the function is not invoked in the step, and AFAICT this is not used elsewhere, so I wonder if that might be a leftover from a WIP state of the PR?
Feel free to correct me if I'm wrong though, I might've missed something

Reply:
It was left over from a WIP version of the PR where the returned artifact was a running VM. I noted in https://github.com/hashicorp/packer-plugin-proxmox/pull/72#issuecomment-2022956177 a desire to have a running VM returned, so I've rerolled with those changes.

Reply:
Ah yeah understood, I missed the resume VM part.

I wonder though, the current workflow appears to work (can't test it as I don't have a proxmox machine on hand), but I would think we can keep the VM running throughout the process if we're not exporting to a template, don't we?
In this case we can change a bit the `Cleanup` method of the `step_start_vm` so it doesn't kill and delete the VM, unless it cannot?
Also to be sure, since the cleanup for the VM will run after `step_finalize_config`, will the VM still exist? I'd assume this will be stopped and removed, won't it?

Reply:
A stop and start of the VM is required if there are VM hardware changes for the VM's Cloud-Init storage and CD-ROM devices (step `step_finalize_config`).

These changes aren't picked up by the VM unless it's completely stopped then started again, they can't be hot-swapped or mounted on an ACPI reboot. This is a limitation on the Proxmox backend.

I've reworked the logic to only stop the VM artifact if there are changes to be done in `step_finalize_config` (if len(changes) > 0) that way if Cloud-Init isn't being configured or CD-ROM devices are being kept, the VM won't restart.